tapeout/src/main/scala/transforms/Generate.scala

```diff
@@ -10,208 +10,115 @@ import firrtl.annotations.AnnotationYamlProtocol._
-      new passes.memlib.ReplSeqMem(),
-      new passes.clocklist.ClockListTransform()
-    ) } else Seq()
-
-    pre ++ enumerate ++ post
   }
 
-  private def getFirstPhaseAnnotations(top: Boolean): AnnotationMap = {
-    if (top) { 
-      //Load annotations from file
-      val annotationArray = annoFile match {
-        case None => Array[Annotation]()
-        case Some(fileName) => {
-          val annotations = new File(fileName)
-          if(annotations.exists) {
-            val annotationsYaml = io.Source.fromFile(annotations).getLines().mkString("\n").parseYaml
-            annotationsYaml.convertTo[Array[Annotation]]
-          } else {
-            Array[Annotation]()
-          }
-        }
-      }
-      // add new annotations
-      AnnotationMap(Seq(
-        passes.memlib.InferReadWriteAnnotation(
-          s"${synTop.get}"
-        ),
-        passes.clocklist.ClockListAnnotation(
-          s"-c:${synTop.get}:-m:${synTop.get}:${listClocks.get}"
-        ),
-        passes.memlib.ReplSeqMemAnnotation(
-          s"-c:${synTop.get}:${seqMemFlags.get}"
-        )
-      ) ++ annotationArray)
-    } else { AnnotationMap(Seq.empty) }
-  }
 
-  private def getSecondPhasePasses: Seq[Transform] = {
-    // always the same for now
+  private def harnessTransforms: Seq[Transform] = {
+    // XXX this is a hack, we really should be checking the masters to see if they are ExtModules
+    val externals = Set(harnessTop.get, synTop.get, "SimSerial")
     Seq(
       new ConvertToExtMod((m) => m.name == synTop.get),
       new RemoveUnusedModules,
-      new RenameModulesAndInstances((m) => AllModules.rename(m))
+      new RenameModulesAndInstances((old) => if (externals contains old) old else (old + "_in" + harnessTop.get))
     )
   }
 
-  // always the same for now
-  private def getSecondPhaseAnnotations: AnnotationMap = AnnotationMap(Seq.empty)
-
   // Top Generation
-  protected def firstPhase(top: Boolean, harness: Boolean): Unit = {
-    require(top || harness, "Must specify either top or harness")
-    firrtl.Driver.compile(
-      input.get,
-      topOutput.getOrElse(output.get),
-      new VerilogCompiler(),
-      Parser.UseInfo,
-      getFirstPhasePasses(top, harness),
-      getFirstPhaseAnnotations(top)
+  protected def executeTop: Unit = {
+
+    optionsManager.firrtlOptions = optionsManager.firrtlOptions.copy(
+      customTransforms = firrtlOptions.customTransforms ++ topTransforms
     )
+
+    firrtl.Driver.execute(optionsManager)
   }
 
   // Harness Generation
-  protected def secondPhase: Unit = {
-    firrtl.Driver.compile(
-      input.get,
-      harnessOutput.getOrElse(output.get),
-      new VerilogCompiler(),
-      Parser.UseInfo,
-      getSecondPhasePasses,
-      getSecondPhaseAnnotations
+  protected def executeHarness: Unit = {
+
+    optionsManager.firrtlOptions = optionsManager.firrtlOptions.copy(
+      customTransforms = harnessTransforms
     )
+
+    firrtl.Driver.execute(optionsManager)
```

Comment:
Is there an issue tracking the double firrtl execution?

Reply:
#37